storage/diemdb/src/lib.rs

```diff
@@ -944,6 +938,85 @@ impl DbReader for DiemDB {
         })
     }
 
+    fn get_event_by_version_with_proof(
+        &self,
+        event_key: &EventKey,
+        event_version: u64,
+        proof_version: u64,
+    ) -> Result<EventByVersionWithProof> {
+        gauged_api("get_event_by_version_with_proof", || {
+            let latest_version = self.get_latest_version()?;
+            ensure!(
+                proof_version <= latest_version,
+                "cannot construct proofs for a version that doesn't exist yet: proof_version: {}, latest_version: {}",
+                proof_version, latest_version,
+            );
+            ensure!(
+                event_version <= proof_version,
+                "event_version {} must be <= proof_version {}",
+                event_version,
+                proof_version,
+            );
+
+            // Get the latest sequence number of an event at or before the
+            // requested event_version.
+            let maybe_seq_num = self
+                .event_store
+                .get_latest_sequence_number(event_version, &event_key)?;
+
+            let (lower_bound_incl, upper_bound_excl) = if let Some(seq_num) = maybe_seq_num {
+                // We need to request the surrounding events (surrounding
+                // as in E_i.version <= event_version < E_{i+1}.version) in order
+                // to prove that there are no intermediate events, i.e.,
+                // E_j, where E_i.version < E_j.version <= event_version.
+                //
+                // This limit also works for the case where `event_version` is
+                // after the latest event, since the upper bound will just be None.
+                let limit = 2;
+
+                let events = self.get_events_with_proof_by_event_key(
+                    &event_key,
+                    seq_num,
+                    Order::Ascending,
+                    limit,
+                    proof_version,
+                )?;
+
+                let mut events_iter = events.into_iter();
+                let lower_bound_incl = events_iter.next();
+                let upper_bound_excl = events_iter.next();
```

Comment:
When the `lower_bound_incl` event is at exactly `event_version`, we don't need to return the upper bound, right?

Reply:
yep, I was just too lazy 😅

Reply:
thinking about this some more, I don't think it would work if there are multiple events of the same kind emitted by the same transaction (therefore at the same version). If we only returned the `lower_bound_incl` event, it would be ambiguous which of the multiple events is the latest one. If we instead want to return all events at the given version for the given event stream, we would probably need to return  `get_transaction_with_proof(lower_bound_version, include_events: true)` (i.e., _all_ events) and the `upper_bound_excl` event, since I'm not sure we can verify that the returned list of events is otherwise complete.

Reply:
Wow, that's a powerful argument, and it makes sense to me.

Reply:
Looks like the way our indices are organized is compatible with this thinking, although one Event in the same stream per version is implicitly the assumption. -- all good to me.